File: apps/marrow/backend/app/schemas/signals.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.insights import TrendPoint


class SignalsMetaResponse(BaseModel):
    days_total: int
    days_usable: int
    warmup: int
    drop_reasons: dict[str, int]
    insufficient_data: bool
    insufficient_reason: Optional[str] = None
    outcome: str
    start: Optional[str] = None
    end: Optional[str] = None


class SignalsModelResponse(BaseModel):
    mae: Optional[float] = None
    baseline_mae: Optional[float] = None
    noise_floor_mae: Optional[float] = None
    noise_sd: Optional[float] = None
    skill: Optional[float] = None
    holdout_rmse: Optional[float] = None
    holdout_r2: Optional[float] = None
    r2_basis: Optional[str] = None
    relearning: bool = False
    relearning_message: Optional[str] = None


class TodayContributionResponse(BaseModel):
    label: str
    detail: Optional[str] = None
    display_value: float
    driver_id: str


class TodayCalibrationPointResponse(BaseModel):
    date: str
    predicted: float
    actual: Optional[float]


class SignalsTodayResponse(BaseModel):
    baseline: Optional[float] = None
    contributions: list[TodayContributionResponse] = Field(default_factory=list)
    predicted: Optional[float] = None
    band_low: Optional[float] = None
    band_high: Optional[float] = None
    band_level: Optional[int] = None
    actual: Optional[float] = None
    residual: Optional[float] = None
    calibration_series: list[TodayCalibrationPointResponse] = Field(default_factory=list)


class DoseBinResponse(BaseModel):
    label: str
    n: int
    mean: Optional[float]


class DayStripsResponse(BaseModel):
    exposed: list[Optional[float]]
    unexposed: list[Optional[float]]


class SignalsDriverResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    feature: str
    label: str
    feature_class: str = Field(alias="class")
    shape: str
    theta_hat: Optional[float]
    ci_low: Optional[float]
    ci_high: Optional[float]
    tier: str
    reason: str
    exposed_days: int
    unexposed_days: int
    exposed_runs: int
    dose_table: list[DoseBinResponse]
    day_strips: DayStripsResponse
    good_direction: Optional[str]
    se_ratio: Optional[float]


class SignalsMirrorResponse(BaseModel):
    feature: str
    label: str
    rho: Optional[float]
    n: int
    reason: str


class UnexplainedEpisodeResponse(BaseModel):
    dates: list[str]
    start_date: str
    end_date: str
    direction: str
    max_abs_residual: float


class TrackerProposalResponse(BaseModel):
    tracker_id: str
    label: str
    days_covered: int


class SignalsUnexplainedResponse(BaseModel):
    unexplained_bad: list[UnexplainedEpisodeResponse]
    unexplained_good: list[UnexplainedEpisodeResponse]
    couldnt_score: list[str]
    relearning: bool
    relearning_message: str
    tracker_proposals: list[TrackerProposalResponse]


class SignalsTrendSeriesResponse(BaseModel):
    key: str
    label: str
    category: str
    points: list[TrendPoint]
    current: Optional[float]
    rolling_avg_7: Optional[float]
    delta_30d: Optional[float]
    good_direction: Optional[str]


class SignalsTrendsResponse(BaseModel):
    series: list[SignalsTrendSeriesResponse]


class SignalsResponse(BaseModel):
    meta: SignalsMetaResponse
    model: SignalsModelResponse
    today: SignalsTodayResponse
    drivers: list[SignalsDriverResponse]
    mirrors: list[SignalsMirrorResponse]
    unexplained: SignalsUnexplainedResponse
    trends: SignalsTrendsResponse

File: apps/marrow/frontend/lib/api/types/signals.ts
/** Hand-written types for GET /api/v1/signals — not health-signals (weather). */

import type { TrendPoint } from './insights'

export type GoodDirection = 'up' | 'down' | null

export interface SignalsMeta {
  days_total: number
  days_usable: number
  warmup: number
  drop_reasons: Record<string, number>
  insufficient_data: boolean
  insufficient_reason: string | null
  outcome: string
  start: string | null
  end: string | null
}

export interface SignalsModel {
  mae: number | null
  baseline_mae: number | null
  noise_floor_mae: number | null
  noise_sd: number | null
  skill: number | null
  holdout_rmse: number | null
  holdout_r2: number | null
  r2_basis: string | null
  relearning: boolean
  relearning_message: string | null
}

export interface TodayContribution {
  label: string
  detail: string | null
  display_value: number
  driver_id: string
}

export interface TodayCalibrationPoint {
  date: string
  predicted: number
  actual: number | null
}

export interface SignalsToday {
  baseline: number | null
  contributions: TodayContribution[]
  predicted: number | null
  band_low: number | null
  band_high: number | null
  band_level: number | null
  actual: number | null
  residual: number | null
  calibration_series: TodayCalibrationPoint[]
}

export interface DoseBin {
  label: string
  n: number
  mean: number | null
}

export interface DayStrips {
  exposed: (number | null)[]
  unexposed: (number | null)[]
}

export interface SignalsDriver {
  feature: string
  label: string
  feature_class: string
  shape: string
  theta_hat: number | null
  ci_low: number | null
  ci_high: number | null
  tier: string
  reason: string
  exposed_days: number
  unexposed_days: number
  exposed_runs: number
  dose_table: DoseBin[]
  day_strips: DayStrips
  good_direction: GoodDirection
  se_ratio: number | null
}

export interface SignalsMirror {
  feature: string
  label: string
  rho: number | null
  n: number
  reason: string
}

export interface UnexplainedEpisode {
  dates: string[]
  start_date: string
  end_date: string
  direction: string
  max_abs_residual: number
}

export interface TrackerProposal {
  tracker_id: string
  label: string
  days_covered: number
}

export interface SignalsUnexplained {
  unexplained_bad: UnexplainedEpisode[]
  unexplained_good: UnexplainedEpisode[]
  couldnt_score: string[]
  relearning: boolean
  relearning_message: string
  tracker_proposals: TrackerProposal[]
}

export interface SignalsTrendSeries {
  key: string
  label: string
  category: string
  points: TrendPoint[]
  current: number | null
  rolling_avg_7: number | null
  delta_30d: number | null
  good_direction: GoodDirection
}

export interface SignalsTrends {
  series: SignalsTrendSeries[]
}

export interface SignalsResponse {
  meta: SignalsMeta
  model: SignalsModel
  today: SignalsToday
  drivers: SignalsDriver[]
  mirrors: SignalsMirror[]
  unexplained: SignalsUnexplained
  trends: SignalsTrends
}

/** Raw driver shape from JSON (`class` alias for feature_class). */
export type SignalsDriverJson = Omit<SignalsDriver, 'feature_class'> & { class: string }
